src/dbnode/encoding/proto/common.go

```diff
@@ -0,0 +1,268 @@
+	// is no reasonm to distinguish between them for the purposes of encoding
+	// and decoding.
+	cNotCustomEncoded customFieldType = iota
+	cSignedInt64
+	cSignedInt32
+	cUnsignedInt64
+	cUnsignedInt32
+	cFloat64
+	cFloat32
+	cBytes
+
+	numCustomTypes = 8
+)
+
+// -1 because iota's are zero-indexed so the highest value will be the number of
+// custom types - 1.
+var numBitsToEncodeCustomType = numBitsRequiredForNumUpToN(numCustomTypes - 1)
+
+const (
+	// Single bit op codes that get encoded into the compressed stream and
+	// inform the iterator / decoder how it should interpret subsequent
+	// bits.
+	opCodeNoMoreDataOrTimeUnitChange = 0
+	opCodeMoreData                   = 1
+
+	opCodeNoMoreData     = 0
+	opCodeTimeUnitChange = 1
+
+	opCodeNoChange = 0
+	opCodeChange   = 1
+
+	opCodeInterpretSubsequentBitsAsLRUIndex          = 0
+	opCodeInterpretSubsequentBitsAsBytesLengthVarInt = 1
+
+	opCodeNoFieldsSetToDefaultProtoMarshal = 0
+	opCodeFieldsSetToDefaultProtoMarshal   = 1
+
+	opCodeIntDeltaPositive = 0
+	opCodeIntDeltaNegative = 1
+
+	opCodeBitsetValueIsNotSet = 0
+	opCodeBitsetValueIsSet    = 1
+)
+
+var (
+	typeOfBytes = reflect.TypeOf(([]byte)(nil))
+
+	// Maps protobuf types to our custom type as described above.
+	mapProtoTypeToCustomFieldType = map[dpb.FieldDescriptorProto_Type]customFieldType{
+		dpb.FieldDescriptorProto_TYPE_DOUBLE: cFloat64,
+		dpb.FieldDescriptorProto_TYPE_FLOAT:  cFloat32,
+
+		dpb.FieldDescriptorProto_TYPE_INT64:    cSignedInt64,
+		dpb.FieldDescriptorProto_TYPE_SFIXED64: cSignedInt64,
+
+		dpb.FieldDescriptorProto_TYPE_UINT64:  cUnsignedInt64,
+		dpb.FieldDescriptorProto_TYPE_FIXED64: cUnsignedInt64,
+
+		dpb.FieldDescriptorProto_TYPE_INT32:    cSignedInt32,
+		dpb.FieldDescriptorProto_TYPE_SFIXED32: cSignedInt32,
+		// Signed because thats how Proto encodes it (can technically have negative
+		// enum values but its not recommended for compression reasons).
+		dpb.FieldDescriptorProto_TYPE_ENUM: cSignedInt32,
+
+		dpb.FieldDescriptorProto_TYPE_UINT32:  cUnsignedInt32,
+		dpb.FieldDescriptorProto_TYPE_FIXED32: cUnsignedInt32,
+
+		dpb.FieldDescriptorProto_TYPE_SINT32: cSignedInt32,
+		dpb.FieldDescriptorProto_TYPE_SINT64: cSignedInt64,
+
+		dpb.FieldDescriptorProto_TYPE_STRING: cBytes,
+		dpb.FieldDescriptorProto_TYPE_BYTES:  cBytes,
+	}
+)
+
+// customFieldState is used to track any required state for encoding / decoding a single
+// field in the encoder / iterator respectively.
+type customFieldState struct {
+	// TODO(rartoul): This could be made more efficient by separating out fields required
+	// for encoding and those required for enumeration, as well as trying to reuse some
+	// of the fields for multiple types to save memory, but its ok for now.
+	fieldNum  int
+	fieldType customFieldType
+
+	// Float state.
+	floatXORState m3tsz.XOREncoder
+
+	// Bytes State.
+	bytesFieldDict         []encoderBytesFieldDictState
+	iteratorBytesFieldDict [][]byte
+
+	// Int state.
+	prevIntBits       uint64
```

Comment:
As discussed lower down, if we can calculate the `NumSig` in the call to `intSigBitsTracker.WriteIntSig(...)` then we can put `prevIntBits` on the `IntSigBitsTracker` and don't need this state here yeah?